File: Intervals/LC_435.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.30

class Solution {
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {
        if (intervals.size() == 1){
            return 0;
        }
        sort(intervals.begin(), intervals.end(), [](const auto& a, const auto& b){
            return a[0] < b[0];  
        });
        int ans = 0;
        vector<vector<int>> result;
        result.push_back(intervals[0]);
        for (int i = 1; i < intervals.size(); i++){
            if (is_overlap(result.back(), intervals[i])){
                ans++;
                // since two intervals are overlapped, we want to make the end
                // to be smaller one, so we could reduce the cut in the future
                result.back()[1] = min(result.back()[1], intervals[i][1]);
                continue;
            }
            else{
                result.push_back(intervals[i]);
            }
        }
        return ans;
    }
private:
    bool is_overlap(vector<int>& l1, vector<int>& l2){
        return l1[1] > l2[0];
    }
};

File: Intervals/LC_435.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.30

class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        intervals.sort()
        ans = 0
        result = [intervals[0]]
        prev_end = intervals[0][1]
        for i in range(1, len(intervals)):
            start, end = intervals[i][0], intervals[i][1]
            if start >= prev_end:
                result.append(intervals[i])
                prev_end = result[-1][1]
            else:
                ans += 1
                prev_end = min(prev_end, end)
        return ans